src/modules/broadcast/broadcast.service.ts

```diff
@@ -1439,62 +1489,124 @@ You can register via: \`\\{unlock\\_link\\}\`
 
             if (message.isPinned) {
               try {
-                await ctx.telegram.pinChatMessage(city.group_id, sentMessage.message_id ?? 0, {
-                  disable_notification: true,
-                });
-              } catch (error) {
-                logs.push(
-                  `☑️ Message delivered to ${city.city_name} (${city.group_id}), but pinning failed: ${error}`,
+                await ctx.telegram.pinChatMessage(
+                  (city.group_id || ctx.chat?.id) ?? 0,
+                  sentMessage?.message_id ?? 0,
+                  {
+                    disable_notification: true,
+                  },
                 );
+              } catch {
+                // logs.push(
+                //   `☑️ Message delivered to ${city.city_name} (${city.group_id}), but pinning failed: ${error}`,
+                // );
               }
             }
 
-            logs.push(`✅ Success: ${city.city_name} (${city.group_id})`);
+            // Escape any commas in the city name for CSV format
+            const escapedCityName = city.city_name.includes(',')
+              ? `"${city.city_name}"`
+              : city.city_name;
+
+            // Add to successful entries array
+            successEntries.push(`${escapedCityName},${city.group_id || ''},Success`);
+
+            await this.saveMessageDetail(broadcastId, sentMessage, city, true);
+          } catch (error) {
+            failureCount++;
+
+            // Escape any commas in the city name for CSV format
+            const escapedCityName = city.city_name.includes(',')
+              ? `"${city.city_name}"`
+              : city.city_name;
+
+            // Add failure entry to CSV with error message
+            const errorMsg = String(error).replace(/"/g, '""');
+            failureEntries.push(`${escapedCityName},${city.group_id || ''},Failed,"${errorMsg}"`);
+
+            // Save the failure in the message detail
+            await this.saveMessageDetail(broadcastId, undefined, city, false);
```

Comment:
We don't need to save the message if its failed to be sent?

Reply:
bad usage of params. The `saveMessageDetail`'s param should be `message_id`, as it does not care about the things included in the entire `Message` object.